syntax = "proto3";

package nuraft_on_the_rocks;

service NuRaftOnTheRocks {
  
  // KV CRUD
  rpc Put (PutRequest) returns (PutResponse);
  rpc Get (GetRequest) returns (GetResponse);
  rpc Delete (DeleteRequest) returns (DeleteResponse);
  rpc Scan (ScanRequest) returns (ScanResponse);

  // Raft Operations
  rpc AddServer (AddServerRequest) returns (AddServerResponse);
  rpc ListServers (ListServersRequest) returns (ListServersResponse);

}

message PutRequest {
  string key = 1;
  string value = 2;
}

message PutResponse {}

message GetRequest {
  string key = 1;
}

message GetResponse {
  string value = 1;
}

message DeleteRequest {
  string key = 1;
}

message DeleteResponse {}

message ScanRequest {
  string start_key = 1;
  string end_key = 2;
}

message ScanResponse {
  repeated KeyValuePair results = 1;
}

message KeyValuePair {
  string key = 1;
  string value = 2;
}

message AddServerRequest {
  string server_id = 1;
  string server_address = 2;
}

message AddServerResponse {}

message ListServersRequest {}

message ServerInfo {
  string server_id = 1;
  string server_address = 2;
  bool is_leader = 3;
}

message ListServersResponse {
  repeated ServerInfo servers = 1;
}
